File: src/main/java/edu/cornell/cs/apl/viaduct/security/solver/ConstantMeetVariableTerm.java
package edu.cornell.cs.apl.viaduct.security.solver;

import com.google.auto.value.AutoValue;

import edu.cornell.cs.apl.viaduct.util.HeytingAlgebra;
import edu.cornell.cs.apl.viaduct.util.dataflow.DataFlowEdge;

/** Meet of a constant element and a variable. */
@AutoValue
public abstract class ConstantMeetVariableTerm<A extends HeytingAlgebra<A>>
    implements LeftHandTerm<A>
{
  public static <A extends HeytingAlgebra<A>> ConstantMeetVariableTerm<A> create(
      A lhs, VariableTerm<A> rhs) {
    return new AutoValue_ConstantMeetVariableTerm<>(lhs, rhs);
  }

  protected abstract A getLhs();

  protected abstract VariableTerm<A> getRhs();

  @Override
  public final ConstraintValue<A> getNode() {
    return getRhs();
  }

  @Override
  public final DataFlowEdge<A> getInEdge() {
    return new PseudocomplementEdge<>(getLhs());
  }
}

File: compiler/src/main/kotlin/io/github/apl_cornell/viaduct/algebra/solver/ConstantMeetVariableTerm.kt
package io.github.apl_cornell.viaduct.algebra.solver

import io.github.apl_cornell.viaduct.algebra.HeytingAlgebra
import io.github.apl_cornell.viaduct.util.dataflow.DataFlowEdge

/** Meet of a constant element and a variable. */
internal data class ConstantMeetVariableTerm<A : HeytingAlgebra<A>>(val lhs: A, val rhs: VariableTerm<A>) :
    LeftHandTerm<A> {
    override fun getValue(solution: ConstraintSolution<A>): A {
        return lhs.meet(rhs.getValue(solution))
    }

    override val node: AtomicTerm<A>
        get() = rhs

    override val inEdge: DataFlowEdge<A>
        get() = ImplyEdge(lhs)
}
